Fix EDConsumerBase to look up consumed module label from EDAlias with product type and instance name as well

makortel opened the pull request:
#### PR description:

This PR resolves https://github.com/cms-sw/cmssw/issues/30937 by making `EDConsumerBase` and `ProductRegistry` to use product type and instance name in addition to the module label when resolving alias-for module label for an EDAlias branch. 

#### PR validation:

Framework unit tests run.

makortel committed:
Specialize ParameterTypeTraits for std::vector<std::pair<std::string, std::vector<T>>>

makortel committed:
Use product type, module label, and instance name to find the module label aliased for by and EDAlias branch

makortel committed:
Add productInstanceName() method to ProductResolverIndexHelper::Matches

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31512/18469

- This PR adds an extra 76KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/src/EDConsumerBase.cc  modified in PR(s): #29553
  - File  FWCore/Integration/test/BuildFile.xml  modified in PR(s): #31129
  - File  FWCore/ParameterSet/interface/ParameterSet.h  modified in PR(s): #31129

cmsbuild commented:
A new Pull Request was created by  @makortel (Matti Kortelainen) for master.

It involves the following packages:

DataFormats/Provenance
FWCore/Framework
FWCore/Integration
FWCore/ParameterSet


@makortel, @smuzaffar, @cmsbuild, @Dr15Jones can you please review it and eventually sign? Thanks.
@rovere, @wddgit this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

makortel commented:
@wddgit @Dr15Jones Could you take a look?

wddgit commented:
I read through it. It looks good to me. It should solve the problem.

cmsbuild commented:
-1

Tested at: 44ce71a670b5157f85b5ea9f10fc143e75214b52

CMSSW: CMSSW_11_2_X_2020-09-18-1100
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-482eab/9436/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests  RelVals  AddOn**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testRecoMETMETProducers had ERRORS


* **RelVals**:

When I ran the RelVals I found an error in the following workflows:
5.1 step1
<pre>runTheMatrix-results/5.1_TTbar+TTbarFS+HARVESTFS/step1_TTbar+TTbarFS+HARVESTFS.log

</pre>135.4 step1
<pre>runTheMatrix-results/135.4_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS/step1_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS.log

</pre>

* **AddOn**:

I found errors in the following addon tests:

cmsDriver.py TTbar_8TeV_TuneCUETP8M1_cfi  --conditions auto:run1_mc --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot Realistic8TeVCollision : FAILED - time: date Sat Sep 19 00:09:52 2020-date Sat Sep 19 00:08:34 2020 s - exit: 16640
cmsDriver.py TTbar_13TeV_TuneCUETP8M1_cfi --conditions auto:run2_mc_l1stage1 --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot NominalCollision2015 --era Run2_25ns : FAILED - time: date Sat Sep 19 00:10:18 2020-date Sat Sep 19 00:08:53 2020 s - exit: 16640
cmsDriver.py TTbar_13TeV_TuneCUETP8M1_cfi --conditions auto:run2_mc --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot NominalCollision2015 --era Run2_2016 : FAILED - time: date Sat Sep 19 00:10:35 2020-date Sat Sep 19 00:08:59 2020 s - exit: 16640

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

makortel committed:
Improve exception message

makortel committed:
Fix consuming EDAlias via edm::View

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31512/18539

- This PR adds an extra 92KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/src/EDConsumerBase.cc  modified in PR(s): #29553
  - File  FWCore/Framework/src/Schedule.cc  modified in PR(s): #31442, #29553

cmsbuild commented:
Pull request #31512 was updated. @makortel, @smuzaffar, @cmsbuild, @Dr15Jones can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: ad4d67e858b3437b79643d37a00f059400c98fa8
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-482eab/9498/summary.html
CMSSW: CMSSW_11_2_X_2020-09-22-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-482eab/9498/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2540471
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2540442
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

makortel commented on DataFormats/Provenance/src/ProductRegistry.cc:
```diff
@@ -558,6 +604,28 @@ namespace edm {
+    if (found == transient_.aliasToOriginal_.end() or aliasFields(*found) != target) {
+      return std::nullopt;
+    }
+    return std::get<Transients::kAliasForModuleLabel>(*found);
```
@Dr15Jones @wddgit 

This approach does not address with the ambiguity if an EDAlias provides e.g. `std::vector<Base>` and `std::vector<Derived>` for the same module label and product instance name, and a consumer consumes `edm::View<Base>`. I was planning to rely on the ambiguity check done also for regular products, but apparently that is done only at the time of `Event::get()` (`consumes()` alone is not sufficient).

The information on `PathsAndConsumesOfModules` will be used for module deletion (#29553; in addition to dependency graph, tracer, etc), I'm wondering if this code should actually report a dependence for each ambiguous case?

Alternatively I could think of changing already `consumes()` to trigger the ambiguity exception (but that may have other consequences).

makortel replied:
In a quick chat with @Dr15Jones we concluded that it could be better to move the ambiguity exception to be thrown at "the `consumes()` time". I'm looking into that next (will be a separate PR).

makortel replied:
The ambiguity exception can not generally be thrown at the beginJob based on consumes information, because the consumption of a product without an explicit process name can only be resolved at the `Event::get()` time.

Given that the ambiguity gets resolved based on products and indices (that are not touched by #29553) leaving this code as it is now could technically be safe for module deletion. However, I think it would be clearest for `PathsAndConsumesOfModules` to report dependencies on all producers of an ambiguous `get()` with `edm::View` of `EDAlias` products.

makortel replied:
> However, I think it would be clearest for `PathsAndConsumesOfModules` to report dependencies on all producers of an ambiguous `get()` with `edm::View` of `EDAlias` products.

The just-added commit adds functionality.

makortel committed:
Make a module consuming an ambiguous EDAlias product with edm::View to depend on all possible producers of that product

While it may be that this is not strictly needed for upcoming deletion
of non-consumed unscheduled modules to work correctly, I think this is
the clearest way for that and other consumers of module consumes
dependence information.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31512/18628

- This PR adds an extra 60KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/src/EDConsumerBase.cc  modified in PR(s): #29553
  - File  FWCore/Framework/src/Schedule.cc  modified in PR(s): #29553
  - File  FWCore/Framework/test/stubs/ToyAnalyzers.cc  modified in PR(s): #29553


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31512/18628/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31512/18628/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31512/18629

- This PR adds an extra 100KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/src/EDConsumerBase.cc  modified in PR(s): #29553
  - File  FWCore/Framework/src/Schedule.cc  modified in PR(s): #29553
  - File  FWCore/Framework/test/stubs/ToyAnalyzers.cc  modified in PR(s): #29553

cmsbuild commented:
Pull request #31512 was updated. @makortel, @smuzaffar, @cmsbuild, @Dr15Jones can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 28cd567882f18ba9193d358206158d227eb31016
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-482eab/9580/summary.html
CMSSW: CMSSW_11_2_X_2020-09-24-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-482eab/9580/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2539438
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2539409
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

makortel commented:
@Dr15Jones @wddgit Could you take a look again?

wddgit commented:
I reviewed it and it looked good to me. +1.

It is probably not worth changing, but sometimes I wonder if the case where a module creates this View ambiguity has ever actually occurred even once in CMSSW. We have some complexity in a few places to deal with it and it might have been simpler to just check for it and make it illegal for a producer to declare it produces collections like that.

makortel commented on DataFormats/Provenance/src/ProductRegistry.cc:
```diff
@@ -457,6 +458,51 @@ namespace edm {
         productTypesConsumed, elementTypesConsumed, containedTypeMap, containedTypeToBaseTypesMap);
+
+    // Add contained types for all EDAliases
+    std::vector<decltype(transient_.aliasToOriginal_)::value_type> elementAliases;
```
The `decltype(...)` should indeed work, and I agree a type alias would more clear. Thanks for the suggestion.

makortel committed:
Use type alias

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31512/18679

- This PR adds an extra 28KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/src/EDConsumerBase.cc  modified in PR(s): #29553
  - File  FWCore/Framework/src/Schedule.cc  modified in PR(s): #29553
  - File  FWCore/Framework/test/stubs/ToyAnalyzers.cc  modified in PR(s): #29553

cmsbuild commented:
Pull request #31512 was updated. @makortel, @smuzaffar, @cmsbuild, @Dr15Jones can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented on DataFormats/Provenance/src/ProductRegistry.cc:
```diff
@@ -457,6 +458,51 @@ namespace edm {
     }
     checkDictionariesOfConsumedTypes(
         productTypesConsumed, elementTypesConsumed, containedTypeMap, containedTypeToBaseTypesMap);
+
```
For clarity, would it be better to move all of this into a separate function?

makortel replied:
Thanks, will do.

Dr15Jones commented on FWCore/Integration/test/PathsAndConsumesOfModulesTestService.cc:
```diff
@@ -0,0 +1,84 @@
+              return desc->moduleLabel() == moduleToCheck.first;
+            });
+        if (found == allModules.end()) {
+          cms::Exception ex("Assert");
```
These seems like a very generic exception name. How about 'TestFailure'?

makortel replied:
Thanks, will change.

Dr15Jones reviewed (commented):
Reviewed all files.

makortel commented:
@cmsbuild, please abort

cmsbuild commented:
Jenkins tests are aborted.

makortel committed:
Move adding element types for aliases to a separate function

makortel committed:
More descriptive exception categories

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31512/18689

- This PR adds an extra 164KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/src/EDConsumerBase.cc  modified in PR(s): #29553
  - File  FWCore/Framework/src/Schedule.cc  modified in PR(s): #29553
  - File  FWCore/Framework/test/stubs/ToyAnalyzers.cc  modified in PR(s): #29553

cmsbuild commented:
Pull request #31512 was updated. @makortel, @smuzaffar, @cmsbuild, @Dr15Jones can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: d61d854069c79bb61c7e23db17beede01fff44f5
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-482eab/9656/summary.html
CMSSW: CMSSW_11_2_X_2020-09-29-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-482eab/9656/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2542225
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2542196
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

makortel commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
+1
